packages/sdk/src/oracle/services/statemachine.ts

```diff
@@ -0,0 +1,108 @@
+import assert from "assert";
+import type { Context, Handlers, Memory, Emit, ContextProps } from "../types/statemachine";
+import uid from "lodash/uniqueId";
+
+export { Context, Handlers, Emit };
+
+/**
+ * StateMachine. This class will be used to handle all change requests by the user, including setting state which
+ * may require triggering fetching data from chain, or initiating transactions that require tracking.
+ *
+ * This class is meant to step through states of a Context object. This object can have variable
+ * parameters and memory. There are several concepts to understand before using this class:
+ *
+ * 1. Handlers - This is the state machine state handlers, it is an objected keyed by each state, with a function handler.
+ * 2. Params - These are common parameters set by the caller, passed into each handler function.
+ * 3. Memory - This is a writing space within the state machine handlers that accumulates through states.
+ * 4. Context - This is the sum of all data needed to transition through the state machine states. It includes params, memory and metadata about the runtime.
+ * 5. Context.done - This is a special property on context.done which represents that the state machine is done transitioning this context.
+ * 6. Context.state = "done" - This is a reserved state on context.state, if set to "done" its the same thing as done = true.
+ * 7. Context.state = "error" - This is a reserved state on context.state, if set to "error" it means the context is done, but also there is an context.error object.
+ * 8. Interval - This is a property on the context which specifies the maximum rate in MS this context will transition, based on the current timestamp.
+ */
+export class StateMachine<P = undefined, M extends Memory = undefined> {
+  private pending: Record<string, Context<P, M>> = {};
+  constructor(public type: string, private handlers: Handlers<P, M>, private emit: Emit<P, M>) {}
```

Comment:
1. if i understand the question right, the benefit of exogenous state here is that its pure data which is useful to the user to be able to inspect, persist or even resume if needed. An example would be tracking a transaction through confirmations. You do want to know when the transaction is mined, but you may also want to know how many confirmations have happened in real time, which is not something you see waiting for a single promise to resolve. theres many ways to skin the cat here, but i thought the best way to do this was separate the stateful context (data layer) from the execution environment.

2. input is not only time, but it is difficult to see without examples.  Input also happens on the creation of a stateful context, this input is set by the user in the `create` function and `params` argument. The only reason why theres a time variable is so that you have some way to delay the machine from spamming an api like "getTransactionHash" as fast as it can waiting for something to happen. Again though its assumed once created (with user input) that the state machine has full control over writing to the object, through the "handlers" passed in the constructor.  Handlers are responsible for interacting with the world ( reading and writing externally) and then advancing state.

I have some simple examples coming in next pr which might make this make some more sense. but yea lmk if my responses are not satisfactory

Reply:
I don't want to get too much into the details on this because it can get really abstract and semantic, but my main confusion came from naming it a state machine, since the state machine is not contained within this class (it is not the only thing modifying the state and the state lives outside of it!).

The primary advantage of a traditional state machine is the separation of input/actions, internal state, and outputs with the main goal being that one can have strong guarantees about the path of the state. I don't think that's the goal here, though.